crates/chain-state/src/deferred_trie.rs

```diff
@@ -0,0 +1,437 @@
+    ///
+    /// - If the async task has completed (`Ready`), returns the cached result.
+    /// - If pending, computes synchronously from stored inputs.
+    ///
+    /// Deadlock is avoided as long as the provided ancestors form a true ancestor chain (a DAG):
+    /// - Each block only waits on its ancestors (blocks on the path to the persisted root)
+    /// - Sibling blocks (forks) are never in each other's ancestor lists
+    /// - A block never waits on its descendants
+    ///
+    /// Given that invariant, circular wait dependencies are impossible.
+    #[instrument(level = "debug", target = "engine::tree::deferred_trie", skip_all)]
+    pub fn wait_cloned(&self) -> ComputedTrieData {
+        let mut state = self.state.lock();
+        match &*state {
+            // If the deferred trie data is ready, return the cached result.
+            DeferredState::Ready(bundle) => {
+                DEFERRED_TRIE_METRICS.deferred_trie_async_ready.increment(1);
+                bundle.clone()
+            }
+            // If the deferred trie data is pending, compute the trie data synchronously and return
+            // the result. This is the fallback path if the async task hasn't completed.
+            DeferredState::Pending(inputs) => {
+                DEFERRED_TRIE_METRICS.deferred_trie_sync_fallback.increment(1);
+                let computed = Self::sort_and_build_trie_input(
+                    &inputs.hashed_state,
+                    &inputs.trie_updates,
+                    inputs.anchor_hash,
+                    &inputs.ancestors,
+                );
+                *state = DeferredState::Ready(computed.clone());
+                computed
+            }
+        }
+    }
+}
+
+/// Sorted trie data computed for an executed block.
+/// These represent the complete set of sorted trie data required to persist
+/// block state for, and generate proofs on top of a block.
+#[derive(Clone, Debug, Default)]
+pub struct ComputedTrieData {
+    /// Sorted hashed post-state produced by execution.
+    pub hashed_state: Arc<HashedPostStateSorted>,
+    /// Sorted trie updates produced by state root computation.
+    pub trie_updates: Arc<TrieUpdatesSorted>,
+    /// Trie input bundled with its anchor hash, if available.
+    pub anchored_trie_input: Option<AnchoredTrieInput>,
+}
+
+impl ComputedTrieData {
+    /// Construct a bundle that includes trie input anchored to a persisted ancestor.
+    pub const fn with_trie_input(
+        hashed_state: Arc<HashedPostStateSorted>,
+        trie_updates: Arc<TrieUpdatesSorted>,
+        anchor_hash: B256,
+        trie_input: Arc<TrieInputSorted>,
+    ) -> Self {
+        Self {
+            hashed_state,
+            trie_updates,
+            anchored_trie_input: Some(AnchoredTrieInput { anchor_hash, trie_input }),
+        }
+    }
+
+    /// Construct a bundle without trie input or anchor information.
+    pub const fn without_trie_input(
```

Comment:
this will likely be used by block builders, so we should elaborate on how this differs from `with_trie_input` and what the trie input anchor is exactly